server/server.go

```diff
@@ -1656,6 +1665,106 @@ func (s *StanServer) start(runningState State) error {
 	return nil
 }
 
+// startChannelGossiping sets up a NATS subscription to a cluster gossip
+// subject to receive messages from the cluster on what channels exist. It also
+// starts a goroutine which periodically publishes the channels this server
+// knows about. When a server discovers it's missing channels, it requests the
+// missing channels from another server and creates them. This should only be
+// called if the server is running in clustered mode.
+func (s *StanServer) startChannelGossiping() error {
+	// Setup sub for receiving channel gossip.
+	gossipInbox := fmt.Sprintf("%s.%s.channels", defaultGossipPrefix, s.opts.ID)
+	_, err := s.ncr.Subscribe(gossipInbox, func(m *nats.Msg) {
+		c := &spb.ChannelGossip{}
+		if err := c.Unmarshal(m.Data); err != nil {
+			s.log.Errorf("Received invalid channel gossip message: %v", err)
+			return
+		}
+		if c.NodeID == s.opts.Clustering.NodeID {
+			// Ignore messages from ourselves.
+			return
+		}
+		s.channels.RLock()
+		count := uint32(len(s.channels.channels))
+		s.channels.RUnlock()
+
+		if count >= c.NumChannels {
+			return
+		}
+
+		// If our count is less, request the list of channels from the other
+		// server and reconcile.
+		resp, err := s.ncr.Request(m.Reply, nil, 5*time.Second)
+		if err != nil {
+			s.log.Errorf("Failed to fetch channels from another server: %v", err)
+			return
+		}
+		channelResp := &spb.ChannelResponse{}
+		if err := channelResp.Unmarshal(resp.Data); err != nil {
+			s.log.Errorf("Received invalid channels response from server: %v", err)
+			return
+		}
+		// Reconcile list of channels.
+		for _, channel := range channelResp.Channels {
+			if s.channels.get(channel) == nil {
+				if _, err := s.channels.createChannel(s, channel); err != nil {
+					s.log.Errorf("Failed to create channel %s while reconciling channels: %v", channel, err)
+				} else {
+					s.log.Debugf("Created channel %s while reconciling channels", channel)
+				}
+			}
+		}
+	})
+	if err != nil {
+		return err
+	}
+
+	// Setup sub for receiving channel requests.
+	reqInbox := fmt.Sprintf("%s.%s.channels.%s", defaultGossipPrefix, s.opts.ID, s.opts.Clustering.NodeID)
+	_, err = s.ncr.Subscribe(reqInbox, func(m *nats.Msg) {
+		s.channels.RLock()
+		channels := make([]string, len(s.channels.channels))
+		i := 0
+		for channel, _ := range s.channels.channels {
+			channels[i] = channel
+			i++
+		}
+		s.channels.RUnlock()
+		resp, err := (&spb.ChannelResponse{Channels: channels}).Marshal()
```

Comment:
You can't do that since the list may be too big to fit in a NATS message. I have done splitting of channels lists for partitioning (check `sendChannelsList()` in `partitions.go`), but the code would need to be made more generic (like accept a connection, subject, list of channels, etc...).